README.md
# BMO

[![Docs Status](https://readthedocs.org/projects/sdss-bmo/badge/?version=latest)](http://sdss-bmo.readthedocs.io/en/latest/)

The [BMO](http://adventuretime.wikia.com/wiki/BMO) acquisition camera controller. Documentation shall go [here](http://sdss-bmo.readthedocs.io/). One day.


CHANGELOG.rst
.. _bmo-changelog:

==========
Change Log
==========

This document records the main changes to the BMO code.


.. _changelog-0.2.6:

0.2.6 (2019-08-18)
------------------

Changed
^^^^^^^
* PyGuide now returns the FWHM, which is output and shown in DS9.


.. _changelog-0.2.5:

0.2.5 (2019-08-12)
------------------

Fixed
^^^^^
* Swap values for ALT and AZ from ``tccPos``.


.. _changelog-0.2.4:

0.2.4 (2019-08-11)
------------------

Changed
^^^^^^^
* Hole coordinates are now saved to the header as ``HOLERA`` and ``HOLEDEC``. ``ALT`` and ``AZ`` now refer to the telescope values from the TCC.


.. _changelog-0.2.3:

0.2.3 (2019-07-12)
------------------

Fixed
^^^^^
* Fixed ``CDELT`` in the Manta WCS which was being written in arcsec instead of degrees.


.. _changelog-0.2.2:

0.2.2 (2018-04-26)
------------------

Changed
^^^^^^^
* Pixel scale, focal scale, and image shape are now defined in the config.
* Changes to `~bmo.devices.manta.MantaExposure`:
    * Compression now uses fpack.
    * Background model parameters are saved in the header.
    * Modified to use ``photutils>=0.4`` and ``astropy>=2.0.4``
    * When loading an image from a FITS image using ``MantaExposure.from_fits``, the background model is applied to reconstruct the raw image.
    * Cleaned code.
* Explicitly closes the image after saving it to disk in case that causes a memory leak.


.. _changelog-0.2.1:

0.2.1 (2017-12-17)
------------------

Fixed
^^^^^
* Added a ``check_connection`` decorator to check the TCC connection before writing.


.. _changelog-0.2.0:

0.2.0 (2017-10-09)
------------------

Added
^^^^^
* BMO now provides a full logging system which also catches Twisted messages.
* A fake Vimba controller which provides at least some basic functionality like connecting a camera.
* Outputs actorkeys keywords for camera devices, exposure state, and controller.
* Finding charts are now, by default, grabbed from platelist.
* Acquisition camera positions on the plates are grabbed from the DB.
* Partially filled up documentation on how to use BMO.
* Last centroid positions are stored and used if ``centre_up`` fails getting centroid position from the DS9 regions.
* Logging now allows to redirect debug, info, and warning messages to the actor. This prevents you from having to do a ``log.debug`` and a ``actor.writeToUsers`` for the same message. Instead, just do ``log.debug('message', actor=actor)``.

Changed
^^^^^^^
* Exposing the cameras is now asynchronous. Both cameras can be exposed at the same time without blocking.
* Background estimation now only happens for the first image in the series. For any subsequent image, the first background is subtracted. If the exposure time changes, the background is recalculated.
* ``bmo ds9 clear`` is now an alias to ``bmo ds9 reset``.

Fixed
^^^^^
* Thanks to the asynchronous exposing, the problem with TCC status timing out during centring up should now be fixed.
* Better handling of TCC disconnections and reconnections.


.. _changelog-0.1.0:

0.1.0 (2017-09-15)
------------------

This is the first release of BMO so this list includes only the most notable features to date.

Added
^^^^^
* Base actor based on `twistedActor <https://github.com/ApachePointObservatory/twistedActor>`_ and TCC device.
* Basic camera control for Allied Vision Manta devices, with camera auto-discovery.
* Display of camera images using DS9 and XPA.
* Command parser based on `click <http://click.pocoo.org/5/>`_. Help commands are broadcasted as responses to the user.
* Basic set of camera control commands (``expose``, ``exptime``, ``reconnect``).
* Loaded cart auto-detection and finding chart display.
* Centroid detection based on `PyGuide <https://github.com/r-owen/PyGuide>`_.
* Centre-up in translation (RA/Dec) and rotation.
* Image background subtraction using `photutils <https://github.com/astropy/photutils>`_.
* Basic astrometry for images.
* Sphinx documentation. Uses `sphinx-click <https://github.com/click-contrib/sphinx-click>`_ to auto-generate available commands.


.. x.y.z (unreleased)
.. ------------------
..
.. A short description
..
.. Added
.. ^^^^^
.. * TBD
..
.. Changed
.. ^^^^^^^
.. * TBD
..
.. Fixed
.. ^^^^^
.. * TBD


python/bmo/logger.py
#!/usr/bin/env python
# encoding: utf-8
#
# logger.py
#
# Created by José Sánchez-Gallego on 17 Sep 2017.


from __future__ import division
from __future__ import print_function
from __future__ import absolute_import

import click
import datetime
import json
import logging
import os
import re
import shutil
import traceback
import sys
import warnings

from logging.handlers import TimedRotatingFileHandler
# from textwrap import TextWrapper

from bmo import config, pathlib

from pygments import highlight
from pygments.lexers import get_lexer_by_name
from pygments.formatters import TerminalFormatter

from twisted.logger import formatEvent, globalLogPublisher


# Adds custom log level for print messages
PRINT = 25
logging.addLevelName(PRINT, 'PRINT')

TWISTED = 12
logging.addLevelName(TWISTED, 'TWISTED')


def print_log_level(self, message, *args, **kws):
    self._log(PRINT, message, args, **kws)


def twisted_log_level(self, message, *args, **kws):
    self._log(TWISTED, message, args, **kws)


logging.Logger._print = print_log_level
logging.Logger._twisted = twisted_log_level


def my_except_hook(type, value, tb):
    """A custom hook for printing tracebacks with colours."""

    tbtext = ''.join(traceback.format_exception(type, value, tb))
    lexer = get_lexer_by_name('pytb', stripall=True)
    formatter = TerminalFormatter()
    sys.stderr.write(highlight(tbtext, lexer, formatter))


def colored_formatter(record):
    """Prints log messages with colours."""

    colours = {'info': ('blue', 'normal'),
               'debug': ('magenta', 'normal'),
               'warning': ('yellow', 'normal'),
               'print': ('green', 'normal'),
               'twisted': ('white', 'bold'),
               'error': ('red', 'bold')}

    levelname = record.levelname.lower()

    if levelname == 'error':
        return

    if levelname.lower() in colours:
        levelname_color = colours[levelname][0]
        bold = True if colours[levelname][1] == 'bold' else False
        header = click.style('[{}]: '.format(levelname.upper()), levelname_color, bold=bold)

    message = '{0}'.format(record.msg)

    warning_category = re.match('^(\w+Warning\:).*', message)
    if warning_category is not None:
        warning_category_colour = click.style(warning_category.groups()[0], 'cyan')
        message = message.replace(warning_category.groups()[0], warning_category_colour)

    sub_level = re.match('(\[.+\]:)(.*)', message)
    if sub_level is not None:
        sub_level_name = click.style(sub_level.groups()[0], 'red')
        message = '{}{}'.format(sub_level_name, ''.join(sub_level.groups()[1:]))

    # if len(message) > 79:
    #     tw = TextWrapper()
    #     tw.width = 79
    #     tw.subsequent_indent = ' ' * (len(record.levelname) + 2)
    #     tw.break_on_hyphens = False
    #     message = '\n'.join(tw.wrap(message))

    sys.__stdout__.write('{}{}\n'.format(header, message))
    sys.__stdout__.flush()

    return


class MyFormatter(logging.Formatter):

    warning_fmp = '%(asctime)s - %(levelname)s: %(message)s [%(origin)s]'
    info_fmt = '%(asctime)s - %(levelname)s - %(message)s [%(funcName)s @ ' + \
        '%(filename)s]'

    ansi_escape = re.compile(r'\x1b[^m]*m')

    def __init__(self, fmt='%(levelname)s - %(message)s [%(funcName)s @ ' +
                 '%(filename)s]'):
        logging.Formatter.__init__(self, fmt, datefmt='%Y-%m-%d %H:%M:%S')

    def format(self, record):

        # Save the original format configured by the user
        # when the logger formatter was instantiated
        format_orig = self._fmt

        # Replace the original format with one customized by logging level
        if record.levelno == logging.DEBUG:
            self._fmt = MyFormatter.info_fmt

        elif record.levelno == logging.getLevelName('PRINT'):
            self._fmt = MyFormatter.info_fmt

        elif record.levelno == logging.getLevelName('TWISTED'):
            self._fmt = MyFormatter.info_fmt

        elif record.levelno == logging.INFO:
            self._fmt = MyFormatter.info_fmt

        elif record.levelno == logging.ERROR:
            self._fmt = MyFormatter.info_fmt

        elif record.levelno == logging.WARNING:
            self._fmt = MyFormatter.warning_fmp

        record.msg = self.ansi_escape.sub('', record.msg)

        # Call the original formatter class to do the grunt work
        result = logging.Formatter.format(self, record)

        # Restore the original format configured by the user
        self._fmt = format_orig

        return result


Logger = logging.getLoggerClass()
fmt = MyFormatter()


class LoggerStdout(object):
    """A pipe for stdout to a logger."""

    def __init__(self, level):
        self.level = level

    def write(self, message):

        if message != '\n':
            self.level(message)

    def flush(self):
        pass


class MyLogger(Logger):
    """This class is used to set up the logging system.

    The main functionality added by this class over the built-in
    logging.Logger class is the ability to keep track of the origin of the
    messages, the ability to enable logging of warnings.warn calls and
    exceptions, and the addition of colorized output and context managers to
    easily capture messages to a file or list.

    """

    INFO = 15

    # The default actor to log to. It is set by the set_actor() method.
    _actor = None

    def save_log(self, path):
        shutil.copyfile(self.log_filename, os.path.expanduser(path))

    def _show_warning(self, *args, **kwargs):

        warning = args[0]
        message = '{0}: {1}'.format(warning.__class__.__name__, args[0])
        mod_path = args[2]

        mod_name = None
        mod_path, ext = os.path.splitext(mod_path)
        for name, mod in sys.modules.items():
            path = os.path.splitext(getattr(mod, '__file__', ''))[0]
            if path == mod_path:
                mod_name = mod.__name__
                break

        if mod_name is not None:
            self.warning(message, extra={'origin': mod_name})
        else:
            self.warning(message)

    def _catch_exceptions(self, exctype, value, tb):
        """Catches all exceptions and logs them."""

        # Now we log it.
        self.error('Uncaught exception', exc_info=(exctype, value, tb))

        # First, we print to stdout with some colouring.
        my_except_hook(exctype, value, tb)

    def _set_defaults(self, name,
                      log_level=logging.INFO,
                      log_file_level=logging.DEBUG,
                      log_file_path='~/',
                      redirect_stdout=True):
        """Reset logger to its initial state."""

        # Remove all previous handlers
        for handler in self.handlers[:]:
            self.removeHandler(handler)

        # Set levels
        self.setLevel(logging.DEBUG)

        # Set up the stdout handler
        self.sh = logging.StreamHandler()
        self.sh.emit = colored_formatter
        self.addHandler(self.sh)

        # Set up the main log file handler if requested (but this might fail if
        # configuration directory or log file is not writeable).

        log_file_path = pathlib.Path(log_file_path).expanduser() / '{}.log'.format(name)
        logdir = log_file_path.parent

        try:
            logdir.mkdir(parents=True, exist_ok=True)

            # If the log file exists, backs it up before creating a new file handler
            if log_file_path.exists():
                strtime = datetime.datetime.utcnow().strftime('%Y-%m-%d_%H:%M:%S')
                shutil.move(str(log_file_path), str(log_file_path) + '.' + strtime)

            self.fh = TimedRotatingFileHandler(str(log_file_path), when='midnight', utc=True)
            self.fh.suffix = '%Y-%m-%d_%H:%M:%S'
        except (IOError, OSError) as ee:
            warnings.warn('log file {0!r} could not be opened for writing: '
                          '{1}'.format(log_file_path, ee), RuntimeWarning)
        else:
            self.fh.setFormatter(fmt)
            self.addHandler(self.fh)
            self.fh.setLevel(log_file_level)

        self.sh.setLevel(log_level)

        self.log_filename = log_file_path
        warnings.showwarning = self._show_warning

        # Redirects all stdout to the logger
        if redirect_stdout:
            sys.stdout = LoggerStdout(self._print)

        # Catches exceptions
        sys.excepthook = self._catch_exceptions

    def set_actor(self, value):
        """Sets the default actor to which to log."""

        self._actor = value

    def debug(self, record, actor=None, **kwargs):
        """Logs a debug message, and writes to the actor users."""

        # If actor=False, does not write to the actor. Otherwise, chooses
        # between the actor argument an the default actor for the logger
        actor = (actor or self._actor) if actor is not False else False

        super(MyLogger, self).debug(record, **kwargs)

        if actor:
            actor.writeToUsers('d', 'text={}'.format(json.dumps(str(record))))

    def info(self, record, actor=None, **kwargs):
        """Logs a info message, and writes to the actor users."""

        actor = (actor or self._actor) if actor is not False else False

        super(MyLogger, self).info(record, **kwargs)

        if actor:
            actor.writeToUsers('i', 'text={}'.format(json.dumps(str(record))))

    def warning(self, record, actor=None, **kwargs):
        """Logs a warning message, and writes to the actor users."""

        actor = (actor or self._actor) if actor is not False else False

        kwargs['extra'] = {'origin': 'actor warning'}
        super(MyLogger, self).warning(record, **kwargs)

        if actor:
            actor.writeToUsers('w', 'text={}'.format(json.dumps(str(record))))


logging.setLoggerClass(MyLogger)
log = logging.getLogger(__name__)
log._set_defaults('bmo',
                  log_level=logging.INFO,
                  log_file_level=logging.DEBUG,
                  log_file_path=config['logging']['logdir'],
                  redirect_stdout=True)


# Creates a twisted observer to redirect messages and failures
def twisted_analyze_event(event):

    text = formatEvent(event)

    if text is None:
        text = ''

    if 'log_failure' in event:
        try:
            traceback = event['log_failure'].getTraceback()
        except Exception:
            traceback = '(UNABLE TO OBTAIN TRACEBACK FROM EVENT)\n'
        text = '\n'.join((text, traceback))
        sys.__stdout__.write(text)
        sys.__stdout__.flush()
        log.exception(text)
    else:
        log._twisted(text)


globalLogPublisher.addObserver(twisted_analyze_event)
